speckle/objects/base.py

```diff
@@ -1,30 +1,73 @@
-from pydantic import BaseModel
+from pydantic import BaseModel, validator
 from pydantic.main import Extra
-from typing import Dict, List, Optional, Any
+from typing import ClassVar, Dict, List, Optional, Any, Type
 from speckle.transports.memory import MemoryTransport
 from speckle.logging.exceptions import SpeckleException
 from speckle.objects.units import get_units_from_string
 
 PRIMITIVES = (int, float, str, bool)
 
 
-class Base(BaseModel):
+class _RegisteringBase(BaseModel):
+    """
+    Private Base model for Speckle types.
+
+    This is an implementation detail, please do not use this outside this module.
+
+    This class provides automatic registration of `speckle_type` into a global,
+    (class level) registry for each subclassing type.
+    The type registry is a base for accurate type based (de)serialization.
+    """
+
+    speckle_type: ClassVar[str]
+    _type_registry: ClassVar[Dict[str, Type["Base"]]] = {}
+
+    class Config:
+        validate_assignment = True
+
+    @classmethod
+    def get_registered_type(cls, speckle_type: str) -> Optional[Type["Base"]]:
+        """Get the registered type from the protected mapping via the `speckle_type`"""
+        return cls._type_registry.get(speckle_type, None)
```

Comment:
I think here we'd want to default to `Base` rather than `None` as if a user recipient receives an object but doesn't have the expected type, deserialisation will throw an unhandled error. I believe we spoke about this before but just to recap, in this situation we def do not want the deserialisation to fail so a plain `Base` will do.

Reply:
Its a fair point, would handle this at the side of the deserialization step. Trying to get a type from a function, that implicitly defaults to a value feels like something that requires explanation. On the other hand, returning an explicit None (documented with the type hint, checked by type checking) makes it self explanatory.

But the bigger picture is, that an unknown type should deserialize to Base.

This raises an iteresting question, where the `speckle_type` is currently stored in a class variable, and there is no way to store the unknown `speckle_type` value on the Base object. This makes it a bit hard to debug. I don't have a solution for that from the top of my head. Should I create a new issue for that?

Reply:
Sure, handling it at deserialisation is a good shout as well. I was thinking it might be more friendly for people getting the type it if it defaults to `Base` as I don't really see a use case where you wouldn't want to check for and handle getting `None`, but I take your point of making it explicit.

Hmm yeah we shouldn't be mutating through deserialising as sending the data back will create a mess for anyone else working with the data. The `speckle_type` def needs to be overwritten on deserialisation when we're falling back to a vanilla `Base`. I would flag a new issue though as I think merging in this PR is a higher priority and solving this prob can be done separately.